chain/chain/src/block_processing_utils.rs

```diff
@@ -21,11 +23,20 @@ pub(crate) struct BlockPreprocessInfo {
     pub(crate) challenges_result: ChallengesResult,
     pub(crate) challenged_blocks: Vec<CryptoHash>,
     pub(crate) provenance: Provenance,
+    /// This field will be set when the apply_chunks has finished.
+    /// This is used to provide a way for caller to wait for the finishing of applying chunks of
+    /// a block
+    pub(crate) apply_chunks_done: Arc<OnceCell<()>>,
+    /// This is used to calculate block processing time metric
+    pub(crate) block_start_processing_time: Instant,
 }
 
 /// Blocks which finished pre-processing and are now being applied asynchronously
 pub(crate) struct BlocksInProcessing {
+    // A map that stores all blocks in processing
     preprocessed_blocks: HashMap<CryptoHash, (Block, BlockPreprocessInfo)>,
+    // A map that indexes all blocks here that need to be caught up by their prev hash
+    blocks_to_catch_up: HashMap<CryptoHash, HashSet<CryptoHash>>,
```

Comment:
I am wondering if this hash map is premature optimization? Can we just 

```rust
    pub(crate) fn has_blocks_to_catch_up(&self, prev_hash: &CryptoHash) -> bool {
        self.preprocessed_blocks
            .iter()
            .any(|(_, (block, _))| block.header().prev_hash() == prev_hash)
    }
```

?

We have limit of `5` on the concurrent block processing, so that is `O(1)`. And we don't have to keep reverse index in sync manualy. Arguably it also makes it more immediately obvious what actually is being computed.